lib/index.js

```diff
@@ -57,32 +57,85 @@ module.exports = (function() {
         return obj === undefined ? defaultValue : obj;
     }
 
-   /**
+		/**
     * Delete unused sheets if needed
     */
     Workbook.prototype.deleteSheet = function(sheetName){
       var self = this;
       var sheet = self.loadSheet(sheetName);
 
-      Object.keys(self.archive.files).forEach(function(key){
-        if (key.indexOf(sheet.id) != -1)
-        delete self.archive.files[key];
-      })
+      var sh = self.workbook.find("sheets/sheet[@sheetId='" + sheet.id + "']");
+      self.workbook.find("sheets").remove(sh);
+			
+      var rel = self.workbookRels.find("Relationship[@Id='" + sh.attrib['r:id'] + "']");
+      self.workbookRels.remove(rel);
+
+      self._rebuild();
+      return self
+    };
 
-      var id = self.sheets.findIndex(function(el){return el.id == sheet.id});
-      self.sheets.splice(id, 1);
+    /**
+    * Clone sheets in current workbook template
+    */
+    Workbook.prototype.copySheet = function(sheetName, copyName){
+      var self = this;
+      var sheet = self.loadSheet(sheetName); //filename, name , id, root
+      var indexS = (self.workbook.findall("sheets/sheet").length+1).toString();
+      var fileName = 'worksheets' + '/' + 'sheet' + indexS + '.xml';
+      var arcName = self.prefix + '/' + fileName;
+
+      self.archive.file(arcName, etree.tostring(sheet.root) );			
+      self.archive.files[arcName].options.binary = true;
+
+      var newSheet = etree.SubElement( self.workbook.find('sheets'), 'sheet' );
+      newSheet.attrib.name = copyName || 'Sheet' + indexS;
+      newSheet.attrib.sheetId = indexS; 
+      newSheet.attrib['r:id'] = 'rId' + indexS;
+
+      var newRel = etree.SubElement(self.workbookRels, 'Relationship');
+      newRel.attrib.Type = 'http://schemas.openxmlformats.org/officeDocument/2006/relationships/worksheet';
+      newRel.attrib.Target = fileName;
+
+      self._rebuild();
+//    TODO: work with "definedNames" 
+//    var defn = etree.SubElement(self.workbook.find('definedNames'), 'definedName');
+//
+      return self
+    };
 
-      self.workbook.find("sheets").delItem(id);
 
-      var el = self.workbook.find("definedNames");
-      if (el){
-        el.delItem(id);
-        el.findall("definedName").forEach(function(item, index){
-          item.attrib.localSheetId = index.toString()
-        })
-      }
+    /**
+    *  Partially rebuild nodes after copy/delete sheets
```

Comment:
Could you explain a bit more what's purpose of `_rebuild` function. 
Looks like it make indexes of elements sequential, but why this is needed? is this Excel requirements, or purely aesthetical. That allow doing safe changes in that function later one.

Reply:
each <sheet> 'r:id' attribute in '\xl\workbook.xml'
must point to correct <Relationship> 'Id' in xl\_rels\workbook.xml.rels
copy an delete break this rule
code simplified and comments added in last commit. I dont like name "_rebuild".
maybe "_restoreRels"  || "_restoreRelationshipIDs" is more convenient